### PR TITLE
[REVIEW PENDING] Zip Challenge: Add new serializers for creating challenge.

### LINKED ISSUE #1036: Add serializers for creating challenge using zip file.

### PR DESCRIPTION
**Pull request by RishabhJain2018**
fixes #1036 @deshraj @taranjeet Please review the PR.

### CODE REVIEW

File: apps/challenges/serializers.py [hunk 1]
```diff
@@ -63,11 +67,47 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
         fields = ('id', 'dataset_split', 'challenge_phase', 'challenge_phase_name', 'dataset_split_name', 'visibility')
 
     def get_dataset_split_name(self, obj):
         return obj.dataset_split.name
 
     def get_challenge_phase_name(self, obj):
         return obj.challenge_phase.name
+
+
+class DatasetSplitSerializer(serializers.ModelSerializer):
```
Comment (deshraj) [inline apps/challenges/serializers.py]: I am guessing that you moved the `DatasetSplitSerializer` here so that all the serializers are in alphabetical order, right?

File: apps/challenges/serializers.py [hunk 2]
```diff
@@ -63,11 +67,47 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+    class Meta:
+        model = Challenge
+        fields = ('id', 'title', 'short_description', 'description', 'terms_and_conditions',
+                  'submission_guidelines', 'start_date', 'end_date', 'creator',
+                  'published', 'enable_forum', 'anonymous_leaderboard', 'is_active',)
+
+
+class ZipFileCreateChallengePhaseSplitSerializer(serializers.ModelSerializer):
+    """
+    Serialize the ChallengePhaseSplits Model and is used to create challenge phase splits
```
Comment (deshraj) [inline apps/challenges/serializers.py]: Replace `ChallengePhaseSplits` with `ChallengePhaseSplit`

File: apps/challenges/serializers.py [hunk 3]
```diff
@@ -63,11 +67,47 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+class LeaderboardSerializer(serializers.ModelSerializer):
+    """
+    Serialize the Leaderboard Model.
+    """
+    class Meta:
+        model = Leaderboard
+        fields = '__all__'
+
+
+class ZipFileCreateChallengeSerializer(ChallengeSerializer):
```
Comment (deshraj) [inline apps/challenges/serializers.py]: Replace `ZipFileCreateChallengeSerializer` with `ZipConfigurationChallengeSerializer`

File: apps/challenges/serializers.py (hunk 1 above)
Comment (RishabhJain2018) [inline apps/challenges/serializers.py]: yes.

**Comment by deshraj** [comment]
Looks good. :)

File: apps/challenges/serializers.py [hunk 4]
```diff
@@ -63,11 +73,47 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+    Serialize the Leaderboard Model.
+    """
+    class Meta:
+        model = Leaderboard
+        fields = '__all__'
+
+
+class ZipConfigurationChallengeSerializer(ChallengeSerializer):
+    """
+    Serialize the Challenge Model and is used to create challenge using zip file uploaded by user.
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Can we rewrite the doc string as "Serializer used for creating challenge through zip file". Same goes for ZipFileCreateChallengePhaseSplitSerializer.

File: apps/challenges/serializers.py [hunk 5]
```diff
@@ -30,6 +31,22 @@ class Meta:
                   'published', 'enable_forum', 'anonymous_leaderboard', 'is_active',)
 
 
+class ChallengeConfigurationSerializer(serializers.ModelSerializer):
+    """
+    Serialize the Challenge Configuration Model.
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Writing doc string is a good practice. I really appreciate that. Just that it will be great if we can keep the model name as `ChallengeConfiguration` (no spaces).

File: apps/challenges/serializers.py [hunk 6]
```diff
@@ -48,13 +65,6 @@ class Meta:
                   'is_public', 'is_active', 'codename')
 
 
-class DatasetSplitSerializer(serializers.ModelSerializer):
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Also there is no need of reordering this serializer here. One of the main reason can be it will actually confuse someone of why this model was involved in this commit. Also in the master, its model `DatasetSplit` is not ordered alphabetically. So in future(may be) we will open an issue where in you can reorder this.

One thought for keeping the serializer as it is it actually tells the history of how the model came into existence. Dont know how much of this is true, but yeah one can simply infer that yeah `Challenge` model was introduced before `ChallengePhase` model.

Comment (RishabhJain2018) [inline apps/challenges/serializers.py]: I agree that it can confuse other people in future. I will reorder it as  previously and yes definitely we can open an issue in future to reorder them.

File: apps/challenges/serializers.py [hunk 7]
```diff
@@ -63,11 +64,55 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
 
     class Meta:
         model = ChallengePhaseSplit
-        fields = '__all__'
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: I am actually not sure why is this line removed?

File: apps/challenges/serializers.py [hunk 8]
```diff
@@ -63,11 +64,55 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
         fields = ('id', 'dataset_split', 'challenge_phase', 'challenge_phase_name', 'dataset_split_name', 'visibility')
 
     def get_dataset_split_name(self, obj):
         return obj.dataset_split.name
 
     def get_challenge_phase_name(self, obj):
         return obj.challenge_phase.name
+
+
+class ChallengeConfigurationSerializer(serializers.ModelSerializer):
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Can we have a some what short name like `ChallengeConfigSerializer` for it?

File: apps/challenges/serializers.py [hunk 9]
```diff
@@ -63,11 +64,55 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+        if context:
+            user = context.get('user')
+            kwargs['data']['user'] = user.pk
+
+    class Meta:
+        model = ChallengeConfiguration
+        fields = ('zip_configuration', 'user',)
+
+
+class LeaderboardSerializer(serializers.ModelSerializer):
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: I think this PR is for adding serializers related to Challenge Config. Hence this should not be here. Can you please check ?

File: apps/challenges/serializers.py [hunk 10]
```diff
@@ -63,11 +64,55 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+    """
+    Serialize the Leaderboard Model.
+    """
+    class Meta:
+        model = Leaderboard
+        fields = '__all__'
+
+
+class ZipConfigurationChallengeSerializer(ChallengeSerializer):
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Can we rename it to `ZipChallengeSerializer` ?

File: apps/challenges/serializers.py [hunk 11]
```diff
@@ -63,11 +64,55 @@ class ChallengePhaseSplitSerializer(serializers.ModelSerializer):
...
+    Serializer used for creating challenge through zip file.
+    """
+    class Meta:
+        model = Challenge
+        fields = ('id', 'title', 'short_description', 'description', 'terms_and_conditions',
+                  'submission_guidelines', 'start_date', 'end_date', 'creator',
+                  'published', 'enable_forum', 'anonymous_leaderboard', 'is_active',)
+
+
+class ZipFileCreateChallengePhaseSplitSerializer(serializers.ModelSerializer):
```
Comment (taranjeet) [inline apps/challenges/serializers.py]: Can we rename it to `ZipChallengePhaseSplitSerializer` ?

File: apps/challenges/serializers.py (hunk 7 above)
Comment (RishabhJain2018) [inline apps/challenges/serializers.py]: Actually, this line is updated in the line just after it. So, there is no use of this line. That's why I removed it. Please check [here](https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/serializers.py#L66).

File: apps/challenges/serializers.py (hunk 9 above)
Comment (RishabhJain2018) [inline apps/challenges/serializers.py]: Actually , this PR is adding all the serializers that are needed during the creation of challenge using zip file and as we are creating the Leaderboard model instance there , so I added the serializer for it. Should i create a separate PR for this?

Comment (taranjeet) [inline apps/challenges/serializers.py]: Ohh, then let it be
